### PR TITLE
sys-cluster/k3s: initial import

### PR DESCRIPTION
**Pull request by xmhd**
As requested by @mcharriere import a slightly outdated sys-cluster/k3s from quark-overlay

I would prefer to improve this ebuild where required before we version bump it to latest stable (1.18 branch).

e.g. there are USE flags for running this 'rootless' which I would prefer to do rather than giving root permissions to things which don't require it, so we should identify things like that so we can have the relevant logic added to infrastructure.

### CODE REVIEW

**Review by varilci — COMMENTED** [review]
It gets complied and installed on a fresh environment successfully. May want to test it's capabilities later by running it with some test stuff though.

**Review by T-10810 — APPROVED** [review]
overlay

>

**Review by T-10810 — APPROVED** [review]

File: sys-cluster/k3s/files/k3s.openrc [hunk 1]
```diff
@@ -0,0 +1,34 @@
+#!/sbin/openrc-run
```
Comment (monsieurp) [inline sys-cluster/k3s/files/k3s.openrc:1]: This entire file gives me nightmares. Please don't merge it as-is.

File: sys-cluster/k3s/files/k3s.openrc [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+SYSTEM_NAME=k3s
+[ "${CMD_K3S}" = "server" ] || SYSTEM_NAME=k3s-${CMD_K3S}
+
+LOG_FILE=/var/log/${SYSTEM_NAME}.log
+
+name="k3s"
+name="${SYSTEM_NAME}"
+command="/usr/bin/k3s"
+command_args="${CMD_K3S} ${K3S_OPTIONS} >>${LOG_FILE} 2>&1"
```
Comment (monsieurp) [inline sys-cluster/k3s/files/k3s.openrc:24]: Why `>>${LOG_FILE}` if `output_log` and `error_log` are set?

File: sys-cluster/k3s/files/k3s.openrc [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+start_pre() {
+    rm -f /tmp/k3s.*
+}
+
+supervisor=supervise-daemon
+
+SYSTEM_NAME=k3s
+[ "${CMD_K3S}" = "server" ] || SYSTEM_NAME=k3s-${CMD_K3S}
```
Comment (monsieurp) [inline sys-cluster/k3s/files/k3s.openrc:17]: Ditto.

File: sys-cluster/k3s/files/k3s.openrc [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+# Copyright 2019 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+depend() {
+    after net-online
+    need net
+}
+
+start_pre() {
+    rm -f /tmp/k3s.*
```
Comment (monsieurp) [inline sys-cluster/k3s/files/k3s.openrc:11]: Why?

File: sys-cluster/k3s/files/k3s.openrc [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+command_args="${CMD_K3S} ${K3S_OPTIONS} >>${LOG_FILE} 2>&1"
+
+output_log="${LOG_FILE}"
+error_log="${LOG_FILE}"
+pidfile="/var/run/${SYSTEM_NAME}.pid"
+respawn_delay=5
+
+set -o allexport
+if [ -f /etc/environment ]; then source /etc/environment; fi
+if [ -f /etc/conf.d/k3s ]; then source /etc/conf.d/k3s; fi
```
Comment (monsieurp) [inline sys-cluster/k3s/files/k3s.openrc:33]: No. OpenRC does it already.

File: sys-cluster/k3s/files/k3s.openrc [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+name="${SYSTEM_NAME}"
+command="/usr/bin/k3s"
+command_args="${CMD_K3S} ${K3S_OPTIONS} >>${LOG_FILE} 2>&1"
+
+output_log="${LOG_FILE}"
+error_log="${LOG_FILE}"
+pidfile="/var/run/${SYSTEM_NAME}.pid"
+respawn_delay=5
+
+set -o allexport
```
Comment (monsieurp) [inline sys-cluster/k3s/files/k3s.openrc:31]: Why?

File: sys-cluster/k3s/k3s-0.9.1.ebuild [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+ARCH=\$(go env GOARCH)
+SUFFIX="-\${ARCH}"
+
+VERSION_CONTAINERD=\$(grep github.com/containerd/containerd \$(dirname \$0)/../go.mod | head -n1 | awk '{print \$4}')
+VERSION_CRICTL=\$(grep github.com/kubernetes-sigs/cri-tools \$(dirname \$0)/../go.mod | head -n1 | awk '{print \$4}')
+EOF
+
+	# Disable go generate, create codegen only
+	sed -i -e 's|go generate|go run pkg/codegen/main.go|' scripts/package-cli
```
Comment (monsieurp) [inline sys-cluster/k3s/k3s-0.9.1.ebuild:107]: `|| die`

File: sys-cluster/k3s/k3s-0.9.1.ebuild [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+RESTRICT="strip mirror"
+
+src_prepare() {
+	default
+
+	pushd src/${EGO_PN}
+
+	# Use out version.sh
+	cat <<EOF > scripts/version.sh
+#!/bin/bash
```
Comment (monsieurp) [inline sys-cluster/k3s/k3s-0.9.1.ebuild:93]: `#!/usr/bin/env bash`

File: sys-cluster/k3s/k3s-0.9.1.ebuild [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+
+DESCRIPTION="Lightweight Kubernetes. 5 less than k8s."
+HOMEPAGE="https://k3s.io"
+SRC_URI="${ARCHIVE_URI}
+	${EGO_VENDOR_URI}"
+LICENSE="Apache-2.0"
+SLOT="0"
+
+DEPEND="
+	>=dev-lang/go-1.12
```
Comment (monsieurp) [inline sys-cluster/k3s/k3s-0.9.1.ebuild:72]: This version of go doesn't exist anymore in the Portage tree, `dev-lang/go` suffices.

File: sys-cluster/k3s/k3s-0.9.1.ebuild [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+	"github.com/theckman/go-flock v0.7.1"
+	"github.com/xeipuuv/gojsonpointer 4e3ac2762d5f"
+	"github.com/xeipuuv/gojsonreference bd5ef7bd5415"
+	"github.com/xeipuuv/gojsonschema 1d523034197f"
+	"go.etcd.io/bbolt v1.3.1-etcd.8 github.com/etcd-io/bbolt"
+	"gopkg.in/mgo.v2 a6b53ec6cb22 github.com/go-mgo/mgo"
+	"k8s.io/utils 3a4a5477acf8 github.com/kubernetes/utils"
+	)
+
+inherit eutils golang-build golang-vcs-snapshot
```
Comment (monsieurp) [inline sys-cluster/k3s/k3s-0.9.1.ebuild:53]: Not sure what the `eutils` is inherited for here.

File: sys-cluster/k3s/k3s-0.9.1.ebuild [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+
+src_compile() {
+	pushd src/${EGO_PN}
+
+	export GOPATH="${T}:$(get_golibdir_gopath)"
+	export GOCACHE="${T}/go-cache"
+	export GOFLAGS="-v -work -x -mod=vendor"
+
+	# dont run ./scripts/download but at least:
+	mkdir -p build/static/charts
```
Comment (monsieurp) [inline sys-cluster/k3s/k3s-0.9.1.ebuild:119]: `|| die`

File: sys-cluster/k3s/k3s-0.9.1.ebuild [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+
+	./scripts/build
+	./scripts/package-cli
```
Comment (monsieurp) [inline sys-cluster/k3s/k3s-0.9.1.ebuild:122]: missing `|| die` to both commands.

File: sys-cluster/k3s/files/k3s.openrc [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+    need net
+}
+
+start_pre() {
+    rm -f /tmp/k3s.*
+}
+
+supervisor=supervise-daemon
+
+SYSTEM_NAME=k3s
```
Comment (monsieurp) [inline sys-cluster/k3s/files/k3s.openrc:16]: This should be set in the conf.d file.

File: sys-cluster/k3s/files/k3s.openrc [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+[ "${CMD_K3S}" = "server" ] || SYSTEM_NAME=k3s-${CMD_K3S}
+
+LOG_FILE=/var/log/${SYSTEM_NAME}.log
```
Comment (monsieurp) [inline sys-cluster/k3s/files/k3s.openrc:19]: Also belongs in the conf.d file.

**Review by monsieurp — CHANGES_REQUESTED** [review]
Those changes can be addressed in a subsequent PR.

**Comment by xmhd** [comment]
closing this as the package has just entered ::gentoo and is being maintained there.

**Pull request closed without merging**